internal/tableland/impl/mesa.go

```diff
@@ -35,150 +36,131 @@ func NewTablelandMesa(
-	ctx context.Context,
-	req tableland.RelayWriteQueryRequest,
-) (tableland.RelayWriteQueryResponse, error) {
+func (t *TablelandMesa) RelayWriteQuery(ctx context.Context, statement string) (tables.Transaction, error) {
 	ctxController := ctx.Value(middlewares.ContextKeyAddress)
 	controller, ok := ctxController.(string)
 	if !ok || controller == "" {
-		return tableland.RelayWriteQueryResponse{}, errors.New("no controller address found in context")
+		return nil, errors.New("no controller address found in context")
 	}
 	ctxChainID := ctx.Value(middlewares.ContextKeyChainID)
 	chainID, ok := ctxChainID.(tableland.ChainID)
 	if !ok {
-		return tableland.RelayWriteQueryResponse{}, errors.New("no chain id found in context")
+		return nil, errors.New("no chain id found in context")
 	}
 	stack, ok := t.chainStacks[chainID]
 	if !ok {
-		return tableland.RelayWriteQueryResponse{}, fmt.Errorf("chain id %d isn't supported in the validator", chainID)
+		return nil, fmt.Errorf("chain id %d isn't supported in the validator", chainID)
 	}
 
 	if !stack.AllowTransactionRelay {
-		return tableland.RelayWriteQueryResponse{},
+		return nil,
 			fmt.Errorf("chain id %d does not suppport relaying of transactions", chainID)
 	}
 
-	mutatingStmts, err := t.parser.ValidateMutatingQuery(req.Statement, chainID)
+	mutatingStmts, err := t.parser.ValidateMutatingQuery(statement, chainID)
 	if err != nil {
-		return tableland.RelayWriteQueryResponse{}, fmt.Errorf("validating query: %s", err)
+		return nil, fmt.Errorf("validating query: %s", err)
 	}
 
 	tableID := mutatingStmts[0].GetTableID()
-	tx, err := stack.Registry.RunSQL(ctx, common.HexToAddress(controller), tableID, req.Statement)
+	tx, err := stack.Registry.RunSQL(ctx, common.HexToAddress(controller), tableID, statement)
 	if err != nil {
-		return tableland.RelayWriteQueryResponse{}, fmt.Errorf("sending tx: %s", err)
+		return nil, fmt.Errorf("sending tx: %s", err)
 	}
 
-	response := tableland.RelayWriteQueryResponse{}
-	response.Transaction.Hash = tx.Hash().String()
-	return response, nil
+	return tx, nil
 }
 
 // RunReadQuery allows the user to run SQL.
-func (t *TablelandMesa) RunReadQuery(
-	ctx context.Context,
-	req tableland.RunReadQueryRequest,
-) (tableland.RunReadQueryResponse, error) {
-	readStmt, err := t.parser.ValidateReadQuery(req.Statement)
+func (t *TablelandMesa) RunReadQuery(ctx context.Context, statement string) (interface{}, error) {
+	readStmt, err := t.parser.ValidateReadQuery(statement)
 	if err != nil {
-		return tableland.RunReadQueryResponse{}, fmt.Errorf("validating query: %s", err)
+		return nil, fmt.Errorf("validating query: %s", err)
 	}
 
 	queryResult, err := t.runSelect(ctx, readStmt)
 	if err != nil {
-		return tableland.RunReadQueryResponse{}, fmt.Errorf("running read statement: %s", err)
+		return nil, fmt.Errorf("running read statement: %s", err)
 	}
-	return tableland.RunReadQueryResponse{Result: queryResult}, nil
+	return queryResult, nil
 }
```

Comment:
@asutula, do you think it may be possible to move the formatting modes at this layer in your next PR? Or would that still be a particular feature of the RPC flavor (REST)?

I'd suspect the latter since formatting things are more related to RPC layers than CoreAPIs. If that's the case, this was the what I was referring to in the tension about duplicate work since not every "API feature" can be generalized.

In any case, I don't know yet what would be the ultimate conclusion. If we end up dropping JSON-RPC, then it might be irrelevant.

Reply:
My personal view is that formatting modes should be out of the Core API. The core API should be agnostic on how its internal representation will be presented to the external world. Maybe we'll have a client that for some reason need a slightly different format output (i know we're aiming for consistency across all clients but sometimes a client has specific needs). 

The way I'd design is by having "presenter" or "formatter" components that would receive the Tableland Core Data Structure and format in any way it's relevant. The clients that need consistency would call the same presenter component. 

Hopefully, that makes sense.

Reply:
Makes sense to me.

Reply:
Yea, what @brunocalza described is what i was planning. Basically, the core api will return `UserRows`, then we'll have the presenter layer (this could just be a function) transform that according to the new output format params provided by the request (both REST and JSON RPC). There is one case where the JSON RPC layer will have some special behavior (when the user asks for `jsonl` output) where it will need to either encode the result as base64 string, or error if the `jsonl` is for more than one object being returned. This setup we're talking about will support that kind of flexibility on a per API basis.

Reply:
I think we're all on the same page here, so going to resolve this convo.